activesupport/lib/active_support/lazy_load_hooks.rb

```diff
@@ -27,19 +27,34 @@ def self.extended(base) # :nodoc:
       base.class_eval do
         @load_hooks = Hash.new { |h, k| h[k] = [] }
         @loaded     = Hash.new { |h, k| h[k] = [] }
+        @only_once  = Hash.new { |h, k| h[k] = [] }
```

Comment:
Hi @rafaelfranca , I have been trying this and it gets tricky. The main issue is that `on_load` and `run_load_hooks` can be called in any given order and both execute hooks. If you `run_load_hooks` for two objects like I do [in this test](https://github.com/rails/rails/pull/30045/files#diff-c636eeda0c27a3588b594d1e0f67b87cR57), when `on_load` gets [executed](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/lazy_load_hooks.rb#L37), you don't check `@load_hooks` so you will need some extra structure to know if the hook was ever executed. Having tried this I find the `@only_once` hash more intention revealing and I like how it separates the `:only_once` logic without interfering with the already existing functionality. What do you think?